waspc/data/Generator/templates/react-app/src/actions/index.ts

```diff
@@ -7,7 +7,10 @@ import {
 } from '@tanstack/react-query'
 import { Query } from '../queries';
 
-export type Action<Input, Output> = (args?: Input) => Promise<Output>;
+export type Action<Input, Output> = 
+  [Input] extends [never] ? 
+  (args?: unknown) => Promise<Output> :
+  (args: Input) => Promise<Output>
```

Comment:
To understand why and what's going on here, we need to establish a couple of things.

### Default input and output types for Operations
Users can specify their operation's `Input` and `Output` types, like this:
```typescript
export const updateTaskIsDone: UpdateTask<Pick<Task, "id">, Task> = async ({ id }, context) => {
  // ...
}
```
We also want to allow them to omit input and output types. Sure, they won't get full-stack type safety, but it helps rapid prototyping:
```typescript
export const updateTaskIsDone: UpdateTask = async (_args, context) => {
  // ...
}
```
The main question is: _What should be the default input and output types?_

Assuming we don't want to use `any` unless absolutely necessary, the correct answer is: `never` for the input, `unknown` for the output. Read https://github.com/wasp-lang/wasp/pull/989#discussion_r1101155001 to understand why.

### Frontend behavior with omitted input types
Assuming the user decides not to specify input and output types:
```typescript
export const updateTaskIsDone: UpdateTask = async (_args, context) => {
  // ...
}
```
We want the client RPC call to either accept nothing or anything. More precisely, all following calls should pass the type checker:
```typescript
updateTaskIsDone() 
updateTaskIsDone(12)
updateTaskIsDone("asdf")
```
To implement such behavior, we need to detect whether an action's `Input` type is `never` and type the argument appropriately:
1. `Input` type is `never` (i.e. the user hasn't specified an input type), the action should take a single optional argument of type `unknown`
2. `Input` type isn't `never` (i.e., the user has specified an input type), the action should simply use the user-specified input type.

### Disabling the distribution of `never`
All that's left to answer is _Why do we wrap the types inside an array?_.

The following type wouldn't work:
```typescript
export type Action<Input, Output> = 
  Input extends never ? 
  (args?: unknown) => Promise<Output> :
  (args: Input) => Promise<Output>

type X = Action<never, number> // X is 'never', not (args?: unknown) => Promise<number>
```
This is because conditional types (in our case, `Action`) distribute over unions (in our case, the empty union `never`). We need a way to prevent the distribution, and that's what the array wrapper is for.

Read https://github.com/microsoft/TypeScript/issues/31751#issuecomment-498526919 for more details.

Reply:
I appreciate the depth of this comment. The TS is still a bit of magic to me that enables it, but I appreciate how much thought went into this! :D